starlette/_utils.py

```diff
@@ -70,16 +74,28 @@ async def __aexit__(self, *args: typing.Any) -> None | bool:
         return None
 
 
-@contextmanager
-def collapse_excgroups() -> typing.Generator[None, None, None]:
+@asynccontextmanager
+async def create_collapsing_task_group() -> typing.AsyncGenerator[anyio.abc.TaskGroup, None]:
     try:
-        yield
-    except BaseException as exc:
-        if has_exceptiongroups:  # pragma: no cover
-            while isinstance(exc, BaseExceptionGroup) and len(exc.exceptions) == 1:
-                exc = exc.exceptions[0]
-
-        raise exc
+        async with anyio.create_task_group() as tg:
+            yield tg
+    except BaseExceptionGroup as excs:
+        if len(excs.exceptions) != 1:
+            raise
+
+        exc = excs.exceptions[0]
+        context = exc.__context__
+        tb = exc.__traceback__
+        cause = exc.__cause__
+        sc = exc.__suppress_context__
+        try:
+            raise exc
```

Comment:
```suggestion
            raise exc from exc.__cause__ or exc.__context__
```

This will prevent weird traceback repetition as in https://github.com/pydantic/logfire/issues/1279#issuecomment-3137505961 and also means that existing `__context__` won't be lost from the traceback and replaced with the exception group as the context instead, although it will become `__cause__` instead.